sw/device/lib/base/hardened.h

```diff
@@ -195,4 +199,271 @@ inline uint32_t launder32(uint32_t val) {
   return val;
 }
 
+/**
+ * Launders the pointer-sized value `val`.
+ *
+ * See `launder32()` for detailed semantics.
+ *
+ * @param val A 32-bit integer to launder.
+ * @return A 32-bit integer which will happen to have the same value as `val` at
+ *         runtime.
+ */
+inline uintptr_t launderw(uintptr_t val) {
+  asm("" : "+r"(val));
+  return val;
+}
+
+/**
+ * Creates a reordering barrier for `val`.
+ *
+ * `barrier32()` takes an argument and discards it, while introducing an
+ * "impure" dependency on that value. This forces the compiler to schedule
+ * instructions such that the intermediate `val` actually appears in a
+ * register. Because it is impure, `barrier32()` operations will not be
+ * reordered past each other or MMIO operations, although they can be reordered
+ * past functions if LTO inlines them.
+ *
+ * Most compilers will try to reorder arithmetic operations in such a way
+ * that they form large basic blocks, since modern microarchitectures can
+ * take advantage of instruction-level parallelism. Unfortunately, this means
+ * that instructions that we want to interleave with other instructions are
+ * likely to get separated; this includes static interleavings,
+ * loop-invariant code motion, and some kinds of unroll-and-jam.
+ *
+ * For example, given
+ *
+ * ```
+ * uint32_t product = 5;
+ *
+ * foo();
+ * product *= product;
+ * foo();
+ * product *= product;
+ * foo();
+ * product *= product;
+ * ```
+ *
+ * A compiler will likely reorder this as
+ *
+ * ```
+ * uint32_t product = 5;
+ *
+ * foo();
+ * foo();
+ * foo();
+ * product *= product;
+ * product *= product;
+ * product *= product;
+ * ```
+ *
+ * The compiler is further entitled to constant-propagate `product`.
+ * `barrier32()` can be used to avoid this:
+ *
+ * ```
+ * uint32_t product = 5;
+ * barrier32(product);
+ *
+ * barrier32(foo());
+ * product *= product;
+ * barrier32(product);
+ *
+ * barrier32(foo());
+ * product *= product;
+ * barrier32(product);
+ *
+ * barrier32(foo());
+ * product *= product;
+ * barrier32(product);
```

Comment:
This example doesn't look right to me (assuming the goal is to ensure that the individual squaring operations are retained). `barrier32` won't stop `product` from being constant propagated to each individual `barrier32(product)` call. In godbolt the generated code loads the constants 5, 25, 625 and 1505 independently: https://godbolt.org/z/8fe9evT3Y. I think we'd need something like a volatile variant of `launder32` to prevent constant propagation here.

Reply:
Nah, laundering the constant `5` is sufficient. This is a bit of a sidebar for the actual example, so I've added a comment saying that it's only necessary because the compiler has infinite information about `5` at that position.